### PR TITLE
Output deprecation message when `sanitize` is called.

### PR DESCRIPTION
**Pull request by kaspth**
cc @chancancode

### CODE REVIEW

**Comment by chancancode** [comment]
:heart: y u so fast

Are the warnings going to show up when running the tests here? If so maybe we should disable that by adding some code in the test helper, otherwise it'll be quite annoying :smile:

**Comment by kaspth** [comment]
It seems I didn't copy over the sanitizer tests :fearful:
...so no.

Should we output deprecation messages on `sanitize_css` and all the options [here](https://github.com/kaspth/rails-deprecated_sanitizer/blob/master/lib/rails/deprecated_sanitizer.rb#L36-L154) as well?

File: lib/rails/deprecated_sanitizer/html-scanner/html/sanitizer.rb [hunk 1]
```diff
@@ -1,10 +1,13 @@
 require 'set'
 require 'cgi'
 require 'active_support/core_ext/module/attribute_accessors'
+require 'active_support/deprecation'
 
 module HTML
   class Sanitizer
     def sanitize(text, options = {})
+      ActiveSupport::Deprecation.warn "This sanitize implementation is deprecated and will be replaced by the Rails Html Sanitizer in Rails 5. Add gem 'rails-html-sanitizer' to your gemfile to upgrade. Read the upgrade guide for Rails 4.2 for more information."
```
Comment (chancancode) [inline lib/rails/deprecated_sanitizer/html-scanner/html/sanitizer.rb]: "This sanitize implementation" only makes sense when you are reading the code here, not so much when people simply upgraded their app and restarted the server (which is really who we are targeting).

Perhaps something like...

```
In Rails 5, the underlying implementation of the `sanitize` helper will be replaced by a __faster__,
__stronger__, __better__ implementation :wink: (no, not really, please rewrite this line!). You can
opt into the new implementation today (and remove this warning) by adding the `rails-html-sanitizer`
gem to your Gemfile. Refer to the ... for the differences between the two implementations.
```

I'll leave it to you to digest and improve this :grin: 

But don't use heredoc! https://github.com/rails/rails/commit/8c52480ba6071a12b97f124d9f21df2c39df3ed3

**Comment by chancancode** [comment]
I don't know enough about the other helpers, so I'll leave the judgement call to you.

Ultimately, it's a balance between three things (roughly in that order):
1. Maximizing the chance of this being seen by the relevant users (who use at least one of the affected methods)
2. Keeping the annoyance level tolerable
3. Ease of implementation

Whether that means we add the warning to all or only a few of those methods, showing it on every call or only the first call, etc... you are probably in a better position to decide that given your expertise :grin:

**Comment by kaspth** [comment]
Thanks, @chancancode!

A message in `sanitize_css` isn't needed, it's called when `sanitize` is.

I'm still unsure of the options though. There will be no replacement for them in Rails 5, and that means they should have a warning, right?

File: lib/rails/deprecated_sanitizer/html-scanner/html/sanitizer.rb (hunk 1 above)
Comment (kaspth) [inline lib/rails/deprecated_sanitizer/html-scanner/html/sanitizer.rb]: Thanks, I've updated the message :heart:

**Comment by chancancode** [comment]
If they don't have a replacement, it makes sense to wran I think. Probably a slightly different message too.

**Comment by kaspth** [comment]
Damn it. This should have been against the `rails:master` branch, not my master branch.

**Comment by chancancode** [comment]
@kaspth sorry :sob: :sob: :sob: :sob: :sob: we discussed this on campfire which resulted in a few things:
1. We are not adding the deprecation warning on `sanitize` – here or in the new gem
2. We will (by that, I mean "you should" :wink:) implement the unsupported options on the new gem with the warning that they no longer do anything, and they can install the deprecated gem to get the old behavior back
3. The new html sanitizer gem will become the default for both new and existing apps on 4.2 afterall :trollface:
4. We'll mention this in the upgrade guides along the lines of "this shouldn't really affect you, but if it does, here is how you get the old behavior back (install old gem)" – something like http://edgeguides.rubyonrails.org/upgrading_ruby_on_rails.html#changes-in-json-handling (3.7.3)

You can read the campfire log for details, sorry about this :sob:

**Comment by rafaelfranca** [comment]
:+1:

**Comment by kaspth** [comment]
1. is done because this is closed.
2. https://github.com/rails/rails-html-sanitizer/pull/24
3. Rewrite the sanitizer README: https://github.com/rails/rails-deprecated_sanitizer/pull/4

**Pull request closed without merging**